src/lib/Simplify.hs

```diff
@@ -258,7 +319,16 @@ simplifyApp f xs =
             naryApp a' (map sink $ toList xs)
         caseExpr <- caseComputingEffs e alts' resultTy
         dropSubst $ simplifyExpr caseExpr
-      _ -> naryApp atom $ toList xs
+      Var v ->
+        lookupAtomName v >>= \case
+          TopFunBound _ (UnspecializedTopFun numSpecializationArgs _) ->
+            case splitAtExact numSpecializationArgs (toList xs) of
+              Just (specializationArgs, runtimeArgs) -> do
+                specialized <- emitSpecializationName v specializationArgs
+                naryApp (Var specialized) runtimeArgs
+              Nothing -> error $ "Unexpected function: " ++ pprint atom
```

Comment:
Shouldn't this error message complain about unsaturated application? It is a function, you just didn't see enough arguments to specialize immediately